tools/shell/linenoise.cpp

```diff
@@ -1764,28 +1762,16 @@ void addErrorHighlighting(uint64_t render_start, uint64_t render_end,
             break;
         case ScanState::IN_SINGLE_QUOTE:
             // single quote - all that will get us out is an unescaped single-quote
-            if (c == '\'') {
-                if (i > 0 && l->buf[i - 1] == '\\') { // escaped quote
-                    i++;
-                    break;
-                } else {
-                    // otherwise revert to standard scan state
-                    state = ScanState::STANDARD;
-                    break;
-                }
+            if (c == '\'' && (i == 0 || l->buf[i - 1] != '\\')) {
+                // revert to standard scan state if not escaped quote
+                state = ScanState::STANDARD;
             }
             break;
         case ScanState::IN_DOUBLE_QUOTE:
             // double quote - all that will get us out is an unescaped quote
-            if (c == '"') {
-                if (i > 0 && l->buf[i - 1] == '\\') { // escaped quote
```

Comment:
What happens for the else case now, when the state is not reset?

Reply:
Suppose the buffer looks like this: `\''`. Previously, when we were at the first `'` character, we incremented `i`. This resulted in skipping the next `'` character for the next iteration which was supposed to end the quote and set the state back to `STANDARD`.

In the else case, the character is part of the string (so either a non-quote character or an escaped quote). So we don't do anything.